File: backend/app/matrix_state_detection.py
from __future__ import annotations

import csv
import math
from pathlib import Path
from statistics import fmean, pstdev

from .validators import detect_csv_dialect_from_file, iter_non_empty_csv_rows


MATRIX_STATE_DETECTION_VERSION = 1
MAX_SAMPLED_CELLS = 96
INTEGER_TOLERANCE = 1e-6


def _inverse_log2(value: float) -> float:
    return math.pow(2, value) - 1


def _inverse_log10(value: float) -> float:
    return math.pow(10, value) - 1


INVERSE_LOG_TRANSFORMS = {
    "natural": (50, math.expm1),
    "2": (100, _inverse_log2),
    "10": (20, _inverse_log10),
}


def _coefficient_of_variation(values: list[float]) -> float | None:
    positive_values = [value for value in values if value > 0 and math.isfinite(value)]
    if len(positive_values) < 2:
        return None
    mean_value = fmean(positive_values)
    if mean_value <= 0:
        return None
    return pstdev(positive_values) / mean_value


def _sample_column_indexes(cell_count: int) -> list[int]:
    if cell_count <= MAX_SAMPLED_CELLS:
        return list(range(cell_count))
    if MAX_SAMPLED_CELLS == 1:
        return [0]
    return sorted(
        {
            round(index * (cell_count - 1) / (MAX_SAMPLED_CELLS - 1))
            for index in range(MAX_SAMPLED_CELLS)
        }
    )


def _rounded_metric(value: float | None) -> float | None:
    return round(value, 6) if value is not None and math.isfinite(value) else None


def classify_matrix_state(
    *,
    sampled_value_count: int,
    integer_like_count: int,
    negative_count: int,
    maximum_value: float,
    linear_column_sums: list[float],
    inverse_log_candidates: dict[str, list[float]],
) -> dict:
    if sampled_value_count <= 0:
        return {
            "detected_state": None,
            "confidence": "low",
            "reasons": ["Not enough numeric values were available to classify the matrix."],
        }

    integer_fraction = integer_like_count / sampled_value_count
    linear_sum_cv = _coefficient_of_variation(linear_column_sums)
    inverse_log_cvs = {
        base: _coefficient_of_variation(column_sums)
        for base, column_sums in inverse_log_candidates.items()
    }
    viable_inverse_log_cvs = {
        base: cv
        for base, cv in inverse_log_cvs.items()
        if cv is not None
    }
    best_inverse_log_base = (
        min(viable_inverse_log_cvs, key=viable_inverse_log_cvs.get)
        if viable_inverse_log_cvs
        else None
    )
    inverse_log_sum_cv = (
        viable_inverse_log_cvs[best_inverse_log_base]
        if best_inverse_log_base is not None
        else None
    )

    metrics = {
        "sampled_values": sampled_value_count,
        "integer_fraction": _rounded_metric(integer_fraction),
        "negative_values": negative_count,
        "maximum_value": _rounded_metric(maximum_value),
        "linear_sum_cv": _rounded_metric(linear_sum_cv),
        "inverse_log_sum_cv": _rounded_metric(inverse_log_sum_cv),
        "inverse_log_base": best_inverse_log_base,
    }

    if negative_count:
        return {
            "detected_state": None,
            "confidence": "low",
            "reasons": [
                "Negative values suggest scaled or centered data.",
                "Raw, normalized, and log-normalized matrices are expected to be non-negative.",
            ],
            "metrics": metrics,
        }

    linear_sums_are_constant = linear_sum_cv is not None and linear_sum_cv <= 0.03
    inverse_sums_are_constant = (
        inverse_log_sum_cv is not None and inverse_log_sum_cv <= 0.03
    )
    inverse_is_clearly_better = (
        inverse_log_sum_cv is not None
        and linear_sum_cv is not None
        and inverse_log_sum_cv <= max(0.01, linear_sum_cv * 0.45)
    )

    if integer_fraction >= 0.999:
        if linear_sums_are_constant:
            return {
                "detected_state": "normalized",
                "confidence": "medium",
                "reasons": [
                    "Values are integer-like, but cell totals are nearly constant.",
                    "This pattern is more consistent with rounded normalized values than raw counts.",
                ],
                "metrics": metrics,
            }
        return {
            "detected_state": "raw",
            "confidence": "high",
            "reasons": [
                "Nearly all sampled values are non-negative integers.",
                "Cell totals vary, which is typical of unnormalized count data.",
            ],
            "metrics": metrics,
        }

    if inverse_sums_are_constant and inverse_is_clearly_better:
        return {
            "detected_state": "log_normalized",
            "confidence": "high",
            "reasons": [
                "Values are non-integer and compressed.",
                (
                    "Reversing the likely log transform produces nearly "
                    "constant cell totals."
                ),
            ],
            "metrics": metrics,
        }

    if linear_sums_are_constant:
        return {
            "detected_state": "normalized",
            "confidence": "high",
            "reasons": [
                "Values are non-integer and cell totals are nearly constant.",
                "This is typical of library-size normalized expression.",
            ],
            "metrics": metrics,
        }

    if maximum_value <= 30:
        return {
            "detected_state": "log_normalized",
            "confidence": "medium",
            "reasons": [
                "Values are non-integer with a compressed non-negative range.",
                "The original normalization method cannot be proven from values alone.",
            ],
            "metrics": metrics,
        }

    return {
        "detected_state": "normalized",
        "confidence": "medium",
        "reasons": [
            "Values are non-integer and remain on an uncompressed scale.",
            "The original normalization method cannot be proven from values alone.",
        ],
        "metrics": metrics,
    }


def detect_matrix_state(source_expression: Path) -> dict:
    dialect = detect_csv_dialect_from_file(source_expression)
    rows = iter_non_empty_csv_rows(source_expression, dialect)
    try:
        header = next(rows)
    except StopIteration:
        return {
            "version": MATRIX_STATE_DETECTION_VERSION,
            "detected_state": None,
            "confidence": "low",
            "reasons": ["The expression matrix is empty."],
        }

    cell_count = max(0, len(header) - 1)
    sampled_indexes = _sample_column_indexes(cell_count)
    linear_sums = [0.0] * len(sampled_indexes)
    inverse_sums = {
        "natural": [0.0] * len(sampled_indexes),
        "2": [0.0] * len(sampled_indexes),
        "10": [0.0] * len(sampled_indexes),
    }
    inverse_sums_available = {base: True for base in inverse_sums}
    sampled_value_count = 0
    integer_like_count = 0
    negative_count = 0
    maximum_value = 0.0

    try:
        for row in rows:
            for sample_position, cell_index in enumerate(sampled_indexes):
                value_index = cell_index + 1
                if value_index >= len(row):
                    continue
                try:
                    value = float(str(row[value_index]).strip())
                except (TypeError, ValueError):
                    continue
                if not math.isfinite(value):
                    continue

                sampled_value_count += 1
                if value < 0:
                    negative_count += 1
                if abs(value - round(value)) <= INTEGER_TOLERANCE:
                    integer_like_count += 1
                maximum_value = max(maximum_value, value)
                linear_sums[sample_position] += value

                for base, (
                    maximum,
                    inverse_function,
                ) in INVERSE_LOG_TRANSFORMS.items():
                    if inverse_sums_available[base] and 0 <= value <= maximum:
                        inverse_sums[base][sample_position] += inverse_function(value)
                    else:
                        inverse_sums_available[base] = False
    except csv.Error as exc:
        return {
            "version": MATRIX_STATE_DETECTION_VERSION,
            "detected_state": None,
            "confidence": "low",
            "reasons": [f"The expression matrix could not be sampled: {exc}"],
        }

    result = classify_matrix_state(
        sampled_value_count=sampled_value_count,
        integer_like_count=integer_like_count,
        negative_count=negative_count,
        maximum_value=maximum_value,
        linear_column_sums=linear_sums,
        inverse_log_candidates={
            base: column_sums
            for base, column_sums in inverse_sums.items()
            if inverse_sums_available[base]
        },
    )
    return {
        "version": MATRIX_STATE_DETECTION_VERSION,
        "sampled_cells": len(sampled_indexes),
        **result,
    }

File: frontend/app/projects/_lib/matrixStateDetection.ts
export type MatrixState = "raw" | "normalized" | "log_normalized";
export type MatrixStateConfidence = "high" | "medium" | "low";

export type MatrixStateDetection = {
  detectedState: MatrixState | null;
  confidence: MatrixStateConfidence;
  reasons: string[];
  metrics?: {
    sampledValues: number;
    integerFraction: number;
    negativeValues: number;
    maximumValue: number;
    linearSumCv: number | null;
    inverseLogSumCv: number | null;
    inverseLogBase: "natural" | "2" | "10" | null;
  };
};

export type ExpressionMatrixInspection = {
  label: string;
  geneCount: number;
  cellCount: number;
  geneNames: string[];
  detection: MatrixStateDetection;
};

const MAX_SAMPLED_CELLS = 96;
const INTEGER_TOLERANCE = 1e-6;

function parseDelimitedLine(line: string, delimiter: string): string[] {
  const fields: string[] = [];
  let value = "";
  let insideQuotes = false;

  for (let index = 0; index < line.length; index += 1) {
    const character = line[index];
    if (character === '"') {
      if (insideQuotes && line[index + 1] === '"') {
        value += '"';
        index += 1;
      } else {
        insideQuotes = !insideQuotes;
      }
    } else if (character === delimiter && !insideQuotes) {
      fields.push(value);
      value = "";
    } else {
      value += character;
    }
  }
  fields.push(value);
  return fields;
}

function countDelimitedFields(line: string, delimiter: string): number {
  return parseDelimitedLine(line, delimiter).length;
}

function detectDelimiter(header: string): string {
  return [",", "\t", ";"].reduce((bestDelimiter, delimiter) =>
    countDelimitedFields(header, delimiter) >
    countDelimitedFields(header, bestDelimiter)
      ? delimiter
      : bestDelimiter,
  );
}

function sampleColumnIndexes(cellCount: number): number[] {
  if (cellCount <= MAX_SAMPLED_CELLS) {
    return Array.from({ length: cellCount }, (_, index) => index);
  }
  const indexes = new Set<number>();
  for (let index = 0; index < MAX_SAMPLED_CELLS; index += 1) {
    indexes.add(
      Math.round((index * (cellCount - 1)) / (MAX_SAMPLED_CELLS - 1)),
    );
  }
  return [...indexes].sort((left, right) => left - right);
}

function coefficientOfVariation(values: number[]): number | null {
  const positiveValues = values.filter(
    (value) => value > 0 && Number.isFinite(value),
  );
  if (positiveValues.length < 2) return null;
  const mean =
    positiveValues.reduce((sum, value) => sum + value, 0) /
    positiveValues.length;
  if (mean <= 0) return null;
  const variance =
    positiveValues.reduce((sum, value) => sum + (value - mean) ** 2, 0) /
    positiveValues.length;
  return Math.sqrt(variance) / mean;
}

function classifyMatrixState({
  sampledValueCount,
  integerLikeCount,
  negativeCount,
  maximumValue,
  linearColumnSums,
  inverseLogCandidates,
}: {
  sampledValueCount: number;
  integerLikeCount: number;
  negativeCount: number;
  maximumValue: number;
  linearColumnSums: number[];
  inverseLogCandidates: Partial<
    Record<"natural" | "2" | "10", number[]>
  >;
}): MatrixStateDetection {
  if (sampledValueCount <= 0) {
    return {
      detectedState: null,
      confidence: "low",
      reasons: ["Not enough numeric values were available to classify the matrix."],
    };
  }

  const integerFraction = integerLikeCount / sampledValueCount;
  const linearSumCv = coefficientOfVariation(linearColumnSums);
  const inverseLogCvs = Object.entries(inverseLogCandidates)
    .map(([base, values]) => ({
      base: base as "natural" | "2" | "10",
      cv: coefficientOfVariation(values),
    }))
    .filter(
      (candidate): candidate is {
        base: "natural" | "2" | "10";
        cv: number;
      } => candidate.cv !== null,
    );
  const bestInverseLog = inverseLogCvs.reduce<
    { base: "natural" | "2" | "10"; cv: number } | null
  >(
    (best, candidate) =>
      best === null || candidate.cv < best.cv ? candidate : best,
    null,
  );
  const inverseLogSumCv = bestInverseLog?.cv ?? null;
  const metrics = {
    sampledValues: sampledValueCount,
    integerFraction,
    negativeValues: negativeCount,
    maximumValue,
    linearSumCv,
    inverseLogSumCv,
    inverseLogBase: bestInverseLog?.base ?? null,
  };

  if (negativeCount > 0) {
    return {
      detectedState: null,
      confidence: "low",
      reasons: [
        "Negative values suggest scaled or centered data.",
        "Choose the original state only if you know how the matrix was prepared.",
      ],
      metrics,
    };
  }

  const linearSumsAreConstant = linearSumCv !== null && linearSumCv <= 0.03;
  const inverseSumsAreConstant =
    inverseLogSumCv !== null && inverseLogSumCv <= 0.03;
  const inverseIsClearlyBetter =
    inverseLogSumCv !== null &&
    linearSumCv !== null &&
    inverseLogSumCv <= Math.max(0.01, linearSumCv * 0.45);

  if (integerFraction >= 0.999) {
    if (linearSumsAreConstant) {
      return {
        detectedState: "normalized",
        confidence: "medium",
        reasons: [
          "Values are integer-like, but cell totals are nearly constant.",
          "This looks more like rounded normalized data than raw counts.",
        ],
        metrics,
      };
    }
    return {
      detectedState: "raw",
      confidence: "high",
      reasons: [
        "Nearly all sampled values are non-negative integers.",
        "Cell totals vary, which is typical of raw counts.",
      ],
      metrics,
    };
  }

  if (inverseSumsAreConstant && inverseIsClearlyBetter) {
    return {
      detectedState: "log_normalized",
      confidence: "high",
      reasons: [
        "Values are non-integer and compressed.",
        "Reversing log1p produces nearly constant cell totals.",
      ],
      metrics,
    };
  }

  if (linearSumsAreConstant) {
    return {
      detectedState: "normalized",
      confidence: "high",
      reasons: [
        "Values are non-integer and cell totals are nearly constant.",
        "This is typical of library-size normalized expression.",
      ],
      metrics,
    };
  }

  if (maximumValue <= 30) {
    return {
      detectedState: "log_normalized",
      confidence: "medium",
      reasons: [
        "Values are non-integer with a compressed non-negative range.",
        "The exact normalization method cannot be proven from values alone.",
      ],
      metrics,
    };
  }

  return {
    detectedState: "normalized",
    confidence: "medium",
    reasons: [
      "Values are non-integer and remain on an uncompressed scale.",
      "The exact normalization method cannot be proven from values alone.",
    ],
    metrics,
  };
}

export async function inspectExpressionMatrix(
  file: File,
): Promise<ExpressionMatrixInspection> {
  const reader = file.stream().getReader();
  const decoder = new TextDecoder();
  let pendingText = "";
  let header: string | null = null;
  let delimiter = ",";
  let geneCount = 0;
  const geneNames: string[] = [];
  let cellCount = 0;
  let sampledIndexes: number[] = [];
  let linearSums: number[] = [];
  let inverseSums: Record<"natural" | "2" | "10", number[]> = {
    natural: [],
    "2": [],
    "10": [],
  };
  const inverseSumsAvailable: Record<"natural" | "2" | "10", boolean> = {
    natural: true,
    "2": true,
    "10": true,
  };
  let sampledValueCount = 0;
  let integerLikeCount = 0;
  let negativeCount = 0;
  let maximumValue = 0;

  const processLine = (rawLine: string) => {
    const line = rawLine.endsWith("\r") ? rawLine.slice(0, -1) : rawLine;
    if (!line.trim()) return;

    if (header === null) {
      header = line.replace(/^\uFEFF/, "");
      delimiter = detectDelimiter(header);
      cellCount = Math.max(
        0,
        countDelimitedFields(header, delimiter) - 1,
      );
      sampledIndexes = sampleColumnIndexes(cellCount);
      linearSums = sampledIndexes.map(() => 0);
      inverseSums = {
        natural: sampledIndexes.map(() => 0),
        "2": sampledIndexes.map(() => 0),
        "10": sampledIndexes.map(() => 0),
      };
      return;
    }

    geneCount += 1;
    const fields = parseDelimitedLine(line, delimiter);
    const geneName = fields[0]?.trim();
    if (geneName) geneNames.push(geneName);
    sampledIndexes.forEach((cellIndex, samplePosition) => {
      const rawValue = fields[cellIndex + 1]?.trim();
      if (!rawValue) return;
      const value = Number(rawValue);
      if (!Number.isFinite(value)) return;

      sampledValueCount += 1;
      if (value < 0) negativeCount += 1;
      if (Math.abs(value - Math.round(value)) <= INTEGER_TOLERANCE) {
        integerLikeCount += 1;
      }
      maximumValue = Math.max(maximumValue, value);
      linearSums[samplePosition] += value;

      if (inverseSumsAvailable.natural && value >= 0 && value <= 50) {
        inverseSums.natural[samplePosition] += Math.expm1(value);
      } else {
        inverseSumsAvailable.natural = false;
      }
      if (inverseSumsAvailable["2"] && value >= 0 && value <= 100) {
        inverseSums["2"][samplePosition] += 2 ** value - 1;
      } else {
        inverseSumsAvailable["2"] = false;
      }
      if (inverseSumsAvailable["10"] && value >= 0 && value <= 20) {
        inverseSums["10"][samplePosition] += 10 ** value - 1;
      } else {
        inverseSumsAvailable["10"] = false;
      }
    });
  };

  while (true) {
    const { value, done } = await reader.read();
    pendingText += decoder.decode(value, { stream: !done });
    const lines = pendingText.split("\n");
    pendingText = lines.pop() ?? "";
    lines.forEach(processLine);
    if (done) break;
  }

  if (pendingText) processLine(pendingText);
  if (header === null) throw new Error("The expression matrix is empty.");

  return {
    label: `${geneCount.toLocaleString()} genes × ${cellCount.toLocaleString()} cells`,
    geneCount,
    cellCount,
    geneNames,
    detection: classifyMatrixState({
      sampledValueCount,
      integerLikeCount,
      negativeCount,
      maximumValue,
      linearColumnSums: linearSums,
      inverseLogCandidates: Object.fromEntries(
        (
          Object.entries(inverseSums) as [
            "natural" | "2" | "10",
            number[],
          ][]
        ).filter(([base]) => inverseSumsAvailable[base]),
      ),
    }),
  };
}
